src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs

```diff
@@ -0,0 +1,471 @@
+  IL_0014:  ldloca.s   V_0
+  IL_0016:  call       ""int System.ReadOnlySpan<byte>.Length.get""
+  IL_001b:  ldloca.s   V_1
+  IL_001d:  call       ""int System.ReadOnlySpan<byte>.Length.get""
+  IL_0022:  ceq
+  IL_0024:  call       ""void System.Console.Write(bool)""
+  IL_0029:  ldnull
+  IL_002a:  call       ""System.ReadOnlySpan<byte> System.ReadOnlySpan<byte>.op_Implicit(byte[])""
+  IL_002f:  stloc.0
+  IL_0030:  ldnull
+  IL_0031:  call       ""System.ReadOnlySpan<byte> System.ReadOnlySpan<byte>.op_Implicit(byte[])""
+  IL_0036:  stloc.1
+  IL_0037:  ldloca.s   V_0
+  IL_0039:  call       ""int System.ReadOnlySpan<byte>.Length.get""
+  IL_003e:  ldloca.s   V_1
+  IL_0040:  call       ""int System.ReadOnlySpan<byte>.Length.get""
+  IL_0045:  ceq
+  IL_0047:  call       ""void System.Console.Write(bool)""
+  IL_004c:  ret
+}");
+        }
+
+        [Fact]
+        [WorkItem(23358, "https://github.com/dotnet/roslyn/issues/23358")]
+        public void EmptyArrayCtor()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System;
+
+class Test
+{
+    public static void Main()
+    {       
+        // inplace inits
+        var s1 = new ReadOnlySpan<sbyte>(new sbyte[]{});
+        var s2 = new ReadOnlySpan<sbyte>(new sbyte[]{});
+
+        Console.Write(s1.Length == s2.Length);
+
+        // make an instance
+        Console.Write(s1.Length == new ReadOnlySpan<sbyte>(new sbyte[]{}).Length);
+    }
+}
+
+", TestOptions.ReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: "TrueTrue", verify: Verification.Passes).VerifyIL("Test.Main", @"
+{
+  // Code size       69 (0x45)
+  .maxstack  2
+  .locals init (System.ReadOnlySpan<sbyte> V_0, //s1
+                System.ReadOnlySpan<sbyte> V_1, //s2
+                System.ReadOnlySpan<sbyte> V_2)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    ""System.ReadOnlySpan<sbyte>""
+  IL_0008:  ldloca.s   V_1
+  IL_000a:  initobj    ""System.ReadOnlySpan<sbyte>""
+  IL_0010:  ldloca.s   V_0
+  IL_0012:  call       ""int System.ReadOnlySpan<sbyte>.Length.get""
+  IL_0017:  ldloca.s   V_1
+  IL_0019:  call       ""int System.ReadOnlySpan<sbyte>.Length.get""
+  IL_001e:  ceq
+  IL_0020:  call       ""void System.Console.Write(bool)""
+  IL_0025:  ldloca.s   V_0
+  IL_0027:  call       ""int System.ReadOnlySpan<sbyte>.Length.get""
+  IL_002c:  ldloca.s   V_2
+  IL_002e:  initobj    ""System.ReadOnlySpan<sbyte>""
+  IL_0034:  ldloc.2
+  IL_0035:  stloc.2
+  IL_0036:  ldloca.s   V_2
+  IL_0038:  call       ""int System.ReadOnlySpan<sbyte>.Length.get""
+  IL_003d:  ceq
+  IL_003f:  call       ""void System.Console.Write(bool)""
+  IL_0044:  ret
+}");
+        }
+
+        [Fact]
+        [WorkItem(23358, "https://github.com/dotnet/roslyn/issues/23358")]
+        public void NotConstArrayCtor()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System;
+
+class Test
+{
+    static int[] arr = new int[]{1, 2, int.Parse(""3""), 4}; 
```

Comment:
I'd like to see some of these tests with `byte[]` as well, since my understanding is that there's different behavior when the element size is 1-byte vs >1 byte.

Reply:
This was actually the only case that did not have a 'byte' scenario. But you are right this is an interesting scenario for `byte[]`

---
In reply to: [171992911](https://github.com/dotnet/roslyn/pull/24621#discussion_r171992911) [](ancestors = 171992911)

Reply:
Whereas int[] will be now rejected due to endianness, the byte[] would be rejected due to having nonconstant elements.

---
In reply to: [172003924](https://github.com/dotnet/roslyn/pull/24621#discussion_r172003924) [](ancestors = 172003924,171992911)